### PR TITLE
base1: Let cockpit.user() return a promise

### PR DESCRIPTION
**Pull request by larsu**

### CODE REVIEW

**Comment by dperpeet** [comment]
unit tests fail: `test-permissions` and `test-user`

File: pkg/base1/cockpit.js [hunk 1]
```diff
@@ -2154,44 +2154,37 @@ function basic_scope(cockpit, jquery) {
...
+                    });
+                })
+                .fail(function(ex) {
+                    dfd.reject(ex);
+                })
+                .always(function() {
+                    dbus.close();
+                });
+        }
+        else {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Should be on the same line as above.

File: pkg/base1/cockpit.js [hunk 2]
```diff
@@ -2154,44 +2154,37 @@ function basic_scope(cockpit, jquery) {
...
-    Object.defineProperty(cockpit, "user", {
-        enumerable: true,
-        get: function user_get() {
-            if (!the_user)
-                the_user = new User();
-            return the_user;
+    cockpit.user = function () {
+        var dfd = cockpit.defer();
+        if (!the_user) {
+            var dbus = cockpit.dbus(null, { "bus": "internal" });
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Due to javascript hoisting insanity, we typically try not to define variables inside an if block, just moving the var above works.

File: pkg/base1/cockpit.js [hunk 3]
```diff
@@ -3771,19 +3764,22 @@ function basic_scope(cockpit, jquery) {
...
-        function user_changed() {
-            var allowed = decide();
-            if (self.allowed !== allowed) {
-                self.allowed = allowed;
-                self.dispatchEvent("changed");
-            }
+        if (self.user) {
+            self.allowed = decide(self.user);
+        }
+        else {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Should be combined with above line.

File: pkg/docker/details.js [hunk 4]
```diff
@@ -91,8 +91,10 @@ define([
 
                     $(commit).find(".container-tag").attr('value', "");
 
-                    var author = cockpit.user["name"] || cockpit.user["user"];
-                    $(commit).find(".container-author").attr('value', author);
+                    cockpit.user().done(function (user) {
+                        var author = user.full_name || cockpit.name;
```
Comment (stefwalter) [inline pkg/docker/details.js]: What is cockpit.name. Did you mean user.name?

File: pkg/shell/credentials.js [hunk 5]
```diff
@@ -30,14 +30,17 @@ define([
...
 
         var watch = null;
         var proc = null;
         var timeout = null;
 
-        refresh();
+        cockpit.user().done(function (user) {
+            self.path = user.home + '/.ssh';
+            refresh();
+        });
```
Comment (stefwalter) [inline pkg/shell/credentials.js]: There's a risk that self.path won't be set by the time various things that need it are called. However this is better than the status quo, where it'll be and stay invalid in that case.

File: pkg/shell/machine-dialogs.js [hunk 6]
```diff
@@ -720,6 +722,9 @@ define([
             error_options = ex;
             keys = credentials.keys_instance();
             $(keys).on("changed", update_keys);
+            cockpit.user().done(function (user) {
+                self.user = user;
+            });
```
Comment (stefwalter) [inline pkg/shell/machine-dialogs.js]: Should we wait to call render() until the above promise completes? ie: .always()?

Comment (larsu) [inline pkg/shell/machine-dialogs.js]: Yeah I agree. Changed it.

File: pkg/shell/credentials.js (hunk 5 above)
Comment (larsu) [inline pkg/shell/credentials.js]: Yeah that's what I figured. I could add checks everywhere if you prefer.

File: pkg/docker/details.js (hunk 4 above)
Comment (larsu) [inline pkg/docker/details.js]: Yep

**Comment by stefwalter** [comment]
This needs a rebase.

**Comment by larsu** [comment]
Thanks, rebased.

File: test/verify/check-multi-machine-key [hunk 7]
```diff
@@ -64,6 +64,7 @@ KEY_IDS_MD5 = [
 class TestMultiMachineKeyAuth(MachineCase):
 
     def load_key(self, name, password):
+        self.browser.wait_text('#content-user-name', 'Administrator')
```
Comment (petervo) [inline test/verify/check-multi-machine-key]: Needs a wait_present first.

Comment (larsu) [inline test/verify/check-multi-machine-key]: Thanks, fixed.

**Pull request closed without merging**